### PR TITLE
querier: include cell stats in page stats

### PR DESCRIPTION
**Pull request by denesb**
We have two mechanism to give visibility into reads having to process many tombstones:
* a warning in the logs, triggered if a read processed more the `tombstone_warn_threshold` dead rows/tombstones
* a trace message, which includes stats of the amount of rows in the page, including the amount of live and dead rows as well as tombstones

This series extends this to also include information on cells, so we have visibility into the case where a read has to process an excessive amount of cell tombstones (mainly because of collections).
A log line is now also logged if the amount of dead cells/tombstones in the page exceeds `tombstone_warn_threshold`. The trace message is also extended to contain cell stats.

The `tombstone_warn_threshold` log lines now receive a 10s rate-limit to avoid excessive log spamming. The rate-limit is separate for the row and cell logs.

Example of the new log line (`tombstone_warn_threshold=10` ):
```
WARN  2024-05-30 07:56:44,979 [shard 0:stmt] querier - Read 98 live cells and 126 dead cells/tombstones for system_schema.scylla_tables <partition-range-scan> (-inf, +inf) (see tombstone_warn_threshold) 
```

Example of the new tracing message:
```
Page stats: 1 partition(s), 0 static row(s) (0 live, 0 dead), 1 clustering row(s) (1 live, 0 dead), 0 range tombstone(s) and 13 cell(s) (1 live, 12 dead) [shard 0] | 2024-05-30 08:13:19.690803 | 127.0.0.1 |           6114 | 127.0.0.1
```

Fixes: https://github.com/scylladb/scylladb/issues/18996

Improvement, not a backport candidate.

### CODE REVIEW

File: mutation/compact_and_expire_result.hh [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+#include <cstdint>
+
+struct compact_and_expire_result {
+    uint64_t live_cells = 0;
+    uint64_t dead_cells = 0;
+    uint32_t collection_tombstones = 0;
+
+    operator bool () const noexcept {
+        return live_cells;
+    }
```
Comment (avikivity) [inline mutation/compact_and_expire_result.hh:20]: I don't think it's intuitive to ask if a compact_and_expire_result is true or false.

It's not completely unrelated (something survived vs nothing survived) but it's also not obvious.

If someone wants to ask whether there are any live cells, they can reference the variable.

File: collection_mutation.cc [hunk 2]
```diff
@@ -166,10 +166,13 @@ collection_mutation_view_description::materialize(const abstract_type& type) con
     return m;
 }
 
-bool collection_mutation_description::compact_and_expire(column_id id, row_tombstone base_tomb, gc_clock::time_point query_time,
+compact_and_expire_result collection_mutation_description::compact_and_expire(column_id id, row_tombstone base_tomb, gc_clock::time_point query_time,
```
Comment (avikivity) [inline collection_mutation.cc:169]: I see why you added operator bool previously. I guess we can keep it, but even the bool return value is on shaky grounds.

**Comment by avikivity** [comment]
Please add a test.

File: collection_mutation.cc (hunk 2 above)
Comment (denesb) [reply collection_mutation.cc:169]: Yes, the operator bool is mainly needed for incremental patching. I can drop it at the end. I don't like it either.

**Comment by denesb** [comment]
v2:
* Rebased.
* Dropped `compact_and_expire::operator bool()`
* Added test.

File: collection_mutation.cc (hunk 2 above)
Comment (denesb) [reply collection_mutation.cc:170]: Dropped in v2.

**Comment by avikivity** [comment]
Idea: cut replica pages based on live+dead cell count, not just memory. This prevents complex queries from dominating reader_concurrency_semaphore.

May annoy users with tiny pages though.

**Pull request merged**